### ISSUE TITLE
[Bug]: Attention Mask Ignored in `transformer_engine` Backend with Packed Sequences (Attention Leakage)

### ISSUE THREAD

**Issue by BlackSamorez** [NVIDIA/Megatron-LM#2357]
## Summary

When training `pretrain_gpt.py` with sequence packing enabled (`--reset-position-ids` and `--reset-attention-mask`) and using the `--transformer-impl transformer_engine` backend, the custom block-diagonal attention mask generated by `GPTDataset` is effectively ignored.

The Transformer Engine (TE) layer defaults to `attn_mask_type='causal'`, which causes it to disregard the `attention_mask` tensor passed during the forward pass. This results in silent attention leakage between unrelated documents within a packed sequence.

## Reproduction Steps
Run `pretrain_gpt.py` with the following combination of flags:
```bash
python pretrain_gpt.py \
    --transformer-impl transformer_engine \
    --reset-position-ids \
    --reset-attention-mask \
````

### 1\. Dataset Does Not Provide `cu_seqlens`

`GPTDataset` generates a dense boolean (or FP8) `attention_mask` tensor to handle document boundaries. It **does not** calculate or return `cu_seqlens` (cumulative sequence lengths) or `PackedSeqParams`.

  * **Reference:** [`[megatron/core/datasets/gpt_dataset.py]`](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/datasets/gpt_dataset.py)

### 2\. TE Defaults to Causal Masking

The Transformer Layer is initialized with an attention mask type of `causal`.

  * [megatron/core/models/gpt/gpt_layer_specs.py](https://github.com/NVIDIA/Megatron-LM/blob/79944055327e4b55f4b07b7cdc028429b851155e/megatron/core/models/gpt/gpt_layer_specs.py#L245)

### 3\. API Contract Violation

According to the [Transformer Engine documentation](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/pytorch.html#transformer_engine.pytorch.DotProductAttention), the `attention_mask` argument in the `forward` pass is conditional:

> Argument attention_mask in the forward call is only used when attn_mask_type includes ‘“padding”’ or “arbitrary”.

Because the configuration remains `'causal'`, TE invokes the underlying kernel (FlashAttention) with `is_causal=True` and no custom mask. This applies a standard lower-triangular mask over the entire packed sequence buffer (`0..args.seq_length`), allowing tokens in Document B to attend to tokens in Document A.

Moreover, if `--reset-position-ids` was used, the documents will have overlapping positions, making the attention confuse them.

## Impact

  * **Correctness:** The autoregressive independence assumption is violated for packed sequences.
  * **Silent Failure:** The model trains without error, but gradients are computed based on invalid context.

## Proposed Solution

The model initialization logic needs to detect if the user has requested a custom mask (via `--reset-attention-mask`) and configure the TE layer accordingly.

**Suggested Logic:**
If `args.reset_attention_mask` is `True`, the `attn_mask_type` passed to `te.pytorch.TransformerLayer` must be forced to `'arbitrary'`. This forces TE to utilize the `attention_mask` tensor provided by the dataset. Or just rewrite the entire `pretrain_gpt.py` up to date with `PackedSeqParams` IDK. Maybe add more asserts on transformer engine arguments.

**Comment by gautham-kollu** [comment]
cc: @erhoo82  for visibility

**Comment by martinjaggi** [comment]
other issue #1878 is probably a duplicate (will be solved by this here)

**Comment by BlackSamorez** [comment]
A bit more info: the `position_ids` resetting doesn't matter unless absolute learned positional embeddings are learned, which is not the case for modern models.

**Comment by sbhavani** [comment]
CC @NVIDIA/megatron-oncall

**Comment by yobibyte** [comment]
I think, this has been addressed in #2461 by @jalbericiola but only for the GRPO training loop. Might be still useful.

**Comment by martinjaggi** [comment]
https://github.com/NVIDIA/Megatron-LM/pull/3386

**Comment by wangcho2k** [comment]
> [#3386](https://github.com/NVIDIA/Megatron-LM/pull/3386)

How about pre-training w/ `--reset-attention-mask`? It seems that #3386 only applies to SFT. Is that correct?

BTW, forcing `attn_mask_type` to `'arbitrary'` effectively disables the `fused` and `flash` DPA backends so I think changing the qkv into thd and calculate cu_seqlens as done in GRPO/SFT is more viable option.

**Comment by Connor-XY** [comment]
The remaining gap appears to be the pretraining path with `--reset-attention-mask`, not only GRPO/SFT. Forcing TE attention to `attn_mask_type='arbitrary'` would make TE consume the dense mask, but it can also disable the fused/flash path. The more scalable fix likely needs the packed/THD representation with `cu_seqlens`, similar to the SFT/GRPO direction, so packed pretraining does not silently train with cross-document leakage.

This is already assigned to @santhnm2. The maintainer decision needed here is whether packed pretraining should be moved to a `PackedSeqParams`/`cu_seqlens` path, or whether Megatron should reject/warn on `transformer_engine + reset_attention_mask` until that path exists.

**Comment by asolergi-nv** [comment]
We are addressing this in https://github.com/NVIDIA/Megatron-LM/pull/5298

**Comment by sbhavani** [comment]
The fix landed in #5298. For TE pretraining, --dataloader-inter-document-masking is now the supported path for inter-document masking. The old --reset-attention-mask dense-mask path should not be used for this case.

Closing this as resolved.
